java/client/src/org/openqa/selenium/remote/internal/HttpClientFactory.java

```diff
@@ -114,18 +119,18 @@ public HttpClient getGridHttpClient(int connection_timeout, int socket_timeout)
         .build();
   }
 
-  private SocketConfig createSocketConfig() {
+  private SocketConfig createSocketConfig(int socketTimeout) {
     return SocketConfig.custom()
         .setSoReuseAddress(true)
-        .setSoTimeout(TIMEOUT_THREE_HOURS)
+        .setSoTimeout(socketTimeout > 0 ? socketTimeout : TIMEOUT_THREE_HOURS)
```

Comment:
I think rather than checking `> 0`, it would make things clearer if you had a static value like `DEFAULT_TIMEOUT = -1` and then use that in `createHttpClient` and change these checks to be `socketTimeout != DEFAULT_TIMEOUT`.  Or something along those lines.